### PR TITLE
Correct alignment of atomic types and (re)add Atomic{I,U}128

### LINKED ISSUE #39590: AtomicU128 uses __atomic_load function on OSX
Last night's nightly build [failed](https://buildbot.rust-lang.org/builders/nightly-dist-rustc-mac/builds/887/steps/compile/logs/stdio) with these logs:

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-L" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/std-a260b5db713b337f.0.o" "-o" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libstd-a260b5db713b337f.dylib" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/std-a260b5db713b337f.metadata.o" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps" "-L" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1-std/release/deps" "-L" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/native/jemalloc/lib" "-L" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/build/compiler_builtins-069104501ba665d8/out" "-L" "/Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/obj/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib" "-l" "System" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/libpanic_unwind-18bf5d50673f1daa.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/libunwind-6b24ec54aa474d14.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/libcollections-4e19d5a43d7fdd2c.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/liballoc-99e0cdfb2e11773a.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/liballoc_jemalloc-7b14c3e63843fe84.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/liblibc-109501e572ed7296.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/libstd_unicode-49cd3c7af2b2f27f.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/librand-664091cbac310259.rlib" "-Wl,-force_load,/tmp/rustc.K4tN3uRshYIw/libcore-f1125930d2c15bcd.rlib" "/tmp/rustc.K4tN3uRshYIw/libcompiler_builtins-594db3b2ae45efeb.rlib" "-l" "pthread" "-l" "c" "-l" "m" "-dynamiclib" "-Wl,-dylib" "-Wl,-install_name,@rpath/libstd-a260b5db713b337f.dylib" "-Wl,-rpath,@loader_path/../lib"
  = note: Undefined symbols for architecture x86_64:
            "___atomic_load", referenced from:
                _$LT$core..sync..atomic..AtomicI128$u20$as$u20$core..fmt..Debug$GT$::fmt::hc3b52e65322d95c5 in libcore-f1125930d2c15bcd.rlib(core-f1125930d2c15bcd.0.o)
                _$LT$core..sync..atomic..AtomicU128$u20$as$u20$core..fmt..Debug$GT$::fmt::h78335460c4fed629 in libcore-f1125930d2c15bcd.rlib(core-f1125930d2c15bcd.0.o)
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: aborting due to previous error

error: Could not compile `std`.
```

Notably it looks like this code:

```rust
#![feature(i128_type, i128)]

use std::sync::atomic::{AtomicU128, Ordering};

#[no_mangle]
pub extern fn foo(a: &AtomicU128) -> u128 {
    a.load(Ordering::SeqCst)
}
```

generates this assembly:

```asm
	.section	__TEXT,__text,regular,pure_instructions
	.globl	_foo
	.p2align	4, 0x90
_foo:
	.cfi_startproc
	pushq	%rbp
Ltmp0:
	.cfi_def_cfa_offset 16
Ltmp1:
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
Ltmp2:
	.cfi_def_cfa_register %rbp
	subq	$16, %rsp
	movq	%rdi, %rax
	leaq	-16(%rbp), %rdx
	movl	$16, %edi
	movl	$5, %ecx
	movq	%rax, %rsi
	callq	___atomic_load
	movq	-16(%rbp), %rax
	movq	-8(%rbp), %rdx
	addq	$16, %rsp
	popq	%rbp
	retq
	.cfi_endproc


.subsections_via_symbols
```

cc @Amanieu and #38959

I'm going to temporarily revert #38959 to fix the nightlies while we figure out a solution

### PR DESCRIPTION
**Pull request by ollie27**
LLVM requires that atomic loads and stores be aligned to at least the size of the type. https://llvm.org/docs/LangRef.html#id193

Increasing the alignment is incompatible with `#[repr(transparent)]` but fortunately that was only recently added but either this PR or a revert of #52149 would have to be backported to 1.29 beta to avoid breakage from that.

Using `#[repr(align)]` is a breaking change though because for some reason you can't put a `#[repr(align)]` struct inside a `#[repr(packed)]` struct. I don't know if there's a way to increase the alignment without that issue.

cc. #32976

Fixes #39590

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I've nominated for beta since a cursory look suggests that's necessary.

File: src/libcore/sync/atomic.rs [hunk 1]
```diff
@@ -124,7 +124,7 @@ pub fn spin_loop_hint() {
 /// [`bool`]: ../../../std/primitive.bool.html
 #[cfg(target_has_atomic = "8")]
 #[stable(feature = "rust1", since = "1.0.0")]
-#[repr(transparent)]
+#[repr(align(1))]
```
Comment (cramertj) [inline src/libcore/sync/atomic.rs]: `align(1)` doesn't do anything here, but `repr(transparent)` does, so I'd opt to use that here.

File: src/libcore/sync/atomic.rs [hunk 2]
```diff
@@ -148,7 +148,9 @@ unsafe impl Sync for AtomicBool {}
 /// This type has the same in-memory representation as a `*mut T`.
 #[cfg(target_has_atomic = "ptr")]
 #[stable(feature = "rust1", since = "1.0.0")]
-#[repr(transparent)]
+#[cfg_attr(target_pointer_width = "16", repr(align(2)))]
```
Comment (cramertj) [inline src/libcore/sync/atomic.rs]: Do we support any platforms where size isn't already equal to alignment for `*mut`? This seems redundant to me, and sacrifices the `#[repr(transparent)]`, which is useful.

File: src/libcore/sync/atomic.rs [hunk 3]
```diff
@@ -1101,7 +1104,7 @@ macro_rules! atomic_int {
         ///
         /// [module-level documentation]: index.html
         #[$stable]
-        #[repr(transparent)]
+        #[repr(align($align))]
```
Comment (cramertj) [inline src/libcore/sync/atomic.rs]: As above, I think that this is already always true with the exception of `u/i128`. I'd prefer a solution which kept `repr(transparent)` intact for other primitives and special-cased the 128-bit types to receive higher-alignment attributes rather than `#[repr(transparent)]`.

Comment (retep998) [reply src/libcore/sync/atomic.rs]: `u64` is only aligned to 4 on 32-bit linux systems.

Comment (cramertj) [reply src/libcore/sync/atomic.rs]: @retep998 Thanks! I think I'd still prefer to special-case and allow `repr(transparent)` on more platforms, but that makes me unsure.

File: src/libcore/sync/atomic.rs (hunk 1 above)
Comment (ollie27) [reply src/libcore/sync/atomic.rs]: Yeah, this was only added for consistency. I don't know if `#[repr(transparent)]` is really appropriate for  `AtomicBool` though because the internal type is actually `u8` and if we wanted to support platforms that don't have byte level atomics it would have to be something bigger.

File: src/libcore/sync/atomic.rs (hunk 2 above)
Comment (ollie27) [reply src/libcore/sync/atomic.rs]: I'm not aware of any but I don't know what platforms Rust might want to support in the future.

File: src/libcore/sync/atomic.rs (hunk 3 above)
Comment (ollie27) [reply src/libcore/sync/atomic.rs]: Yeah, that's the issue, it's not clear which types can be `repr(transparent)`. That's why I just added `repr(align)` to all types to avoid the issue. We presumably definitely don't want the same type to have `repr(transparent)` on some platforms but not others so I don't really know what else to do here.

Comment (cramertj) [reply src/libcore/sync/atomic.rs]: I'm not sure either. Ping @rust-lang/libs for your thoughts? This is a breaking (but necessary) change to the representation guarantees of several atomics-- it'd be nice to keep the `repr(transparent)` where possible, but I agree that having it be platform-dependent is pretty dodgy.

Comment (alexcrichton) [reply src/libcore/sync/atomic.rs]: I'd personally be in favor of removing `#[repr(transparent)]` on the atomic types. I don't recall the motivation for adding them, though, so removing them to me is basically the more conservative thing to do (which is good for maintenance!)

Comment (sfackler) [reply src/libcore/sync/atomic.rs]: I did recently come across a need for atomic loads at specific addresses, which `#[repr(transparent)]` is a way to enable (if not the best way). Specifically the Linux perf_events interface uses a memory mapped ring buffer to transfer sample information between the kernel and user. The head and tail offsets corresponding to the "live" parts of the buffer are at specific offsets in the buffer, and since the kernel can be concurrently adjusting them you'd want to perform atomic loads. We could stabilize `std::ptr::atomic_load(*const u64, Ordering) -> u64` but without that the other option is to construct a `&AtomicU64` at the right address. This only works if `AtomicU64` has the same layout as u64, though.

Comment (alexcrichton) [reply src/libcore/sync/atomic.rs]: @sfackler I think though that the purpose here is that `atomic_load(u64)` for Linux only works on x86_64, not on x86 where it may not be aligned, right?

Comment (sfackler) [reply src/libcore/sync/atomic.rs]: Right, the alignment requirements are higher so `#[repr(transparent)]` may not be the best way to get to the functionality I'm looking for, but there is *some* use to there being representational guarantees about these types in the absence of other APIs.

Comment (jethrogb) [reply src/libcore/sync/atomic.rs]: @sfackler I'd think you can still construct a `*const AtomicU64` at the right address without `repr(transparent)`?

Comment (retep998) [reply src/libcore/sync/atomic.rs]: As long as the layout is the same, then `*const AtomicU64` to arbitrary memory is still fine. Where `repr(transparent)` is needed is specifically for compatibility in function calling convention ABIs because two types having the same layout may not be passed the same.

**Comment by nikomatsakis** [comment]
I am not clear on whether this is @rust-lang/compiler, @rust-lang/libs, or even @rust-lang/lang, but I do agree we ought to backport if we decide to accept this PR.  =) I'm going to cc @rust-lang/wg-unsafe-code-guidelines as well well I'm going crazy because it seems like the sort of thing that is related to unsafe code. 

It seems like the real question at play here is:

- Do we guarantee `#[repr(transparent)]` for atomics *or* do we instead guarantee some kind of alignment requirements?

We sort of can't do both because they may be in conflict?

Is there a conservative decision we can make in the meantime which we can backport?

**Comment by nikomatsakis** [comment]
My feeling btw is that the ultimate decision here probably requires at *least* an FCP and might benefit from an RFC (since there may be subtle constraints at play we are not fully considering) — but if we can take some immediate steps (e.g., removing `#[repr(transparent)]`) that leave room for us to make the decision later, that would be good.

**Comment by joshtriplett** [comment]
For at least u16/u32/u64, it's not clear to me why we can't have both `#[repr(transparent)]` and `#[repr(align(size-of-the-type))]`.

**Comment by RalfJung** [comment]
Removing `repr(transparent)` would be a good step I think. I would expect it to be rare anyway that this is required. This only makes a difference when passing things *by value* to FFI, right?

> Do we guarantee #[repr(transparent)] for atomics or do we instead guarantee some kind of alignment requirements?

Given that LLVM mandates alignment requirements, our hands are bound.

File: src/libcore/sync/atomic.rs (hunk 3 above)
Comment (sfackler) [reply src/libcore/sync/atomic.rs]: My understanding is that `repr(transparent)` is the only thing that guarantees that the layout *is* the same - do we have a separate guarantee that single field structs have the same layout as their element?

**Comment by rkruppe** [comment]
> Do we guarantee #[repr(transparent)] for atomics or do we instead guarantee some kind of alignment requirements?

`repr(transparent)` is a bit of a red herring here, since people only need *memory layout* compatibility, i.e., the ability to cast `*const u32` to `*const AtomicU32` or the like, not the calling conventions stuff that `repr(transparent)` actually adds. As I see it, the alternatives are actually:

- force natural alignment of atomic types (placing some extra burden on people who type-pun them in unsafe code)
- allow atomic types that are not naturally aligned, possibly ruling out some implementation strategies

The latter seems more important to me, but are there any platforms where u64 is not naturally aligned in the ABI, yet 64 bit atomics are supported and require natural alignment? Normally, 64 bit CPUs make u64 naturally aligned while 32 bit CPUs emulate in a library 64 bit atomics in software without corresponding alignment requirements. I suppose a 32 bit platform with double-word CAS might be relevant, but is that even a thing and do we care about it?

File: src/libcore/sync/atomic.rs (hunk 3 above)
Comment (RalfJung) [reply src/libcore/sync/atomic.rs]: When not talking about passing things around in functions by value, wouldn't that be `repr(C)`?

**Comment by rkruppe** [comment]
Oh, I missed that LLVM mandates this. Then I guess it's settled. Note that this has more repercussions than just removing `repr(transparent)` &ndash; with or without the attribute, the heightened alignment requirement means that one can't safely cast any `&u64` to `&AtomicU64` and operate on it, unless `align_of::<u64>() >= 8`. Well, it's really only a problem on platforms where u64 is not naturally aligned yet `AtomicU64` exists, which as I wrote in an earlier comment think is very uncommon.

File: src/libcore/sync/atomic.rs (hunk 3 above)
Comment (rkruppe) [reply src/libcore/sync/atomic.rs]: Yes, `repr(C)` would be one way to achieve layout compatibility. However, since this is in libstd, we can also just *state* that it's layout compatible and rely on the fact that rustc does newtype unwrapping. Furthermore, I am convinced we should guarantee layout compatibility of `repr(Rust)` newtypes anyway.

**Comment by pietroalbini** [comment]
Ping! This PR is beta-nominated, and if we want it to reach 1.29 stable we need to merge this in less than a week. What's the progress?

**Comment by alexcrichton** [comment]
Thanks for the ping @pietroalbini, I've submitted to https://github.com/rust-lang/rust/pull/53978 to temporarily revert https://github.com/rust-lang/rust/pull/52149 on beta-only as it should give us some more time to discuss this PR

**Comment by alexcrichton** [comment]
I've also sent https://github.com/rust-lang/rust/pull/53979 for the master branch

**Comment by eddyb** [comment]
@alexcrichton reminder that we have this weird code in `rustc_codegen_llvm`:
https://github.com/rust-lang/rust/blob/b24330fb7d9e89d63eb03d81fe577172aad49525/src/librustc_codegen_llvm/builder.rs#L489-L492
https://github.com/rust-lang/rust/blob/b24330fb7d9e89d63eb03d81fe577172aad49525/src/librustc_codegen_llvm/builder.rs#L565-L567

Maybe the thing to do here is use natural alignment? (i.e. the size of the integer/pointer type)

**Comment by rkruppe** [comment]
Yes, since LLVM requires natural alignment on atomic loads/stores, we should set the alignment to the size of the type.

**Comment by nagisa** [comment]
There are more requirements to alignment than just specifying the size-of-the-type on the load/store. We should just use the "abi" alignment in that snippet of the code and ICE if "abi" alignment (i.e. the `#[align]`) is smaller than the size of the type.

That being said, this would involve calling the `atomic_` intrinsics with `T = Atomic*` and not `T = {primitive}`, because the integers contained within the wrappers do not inherit the alignment attribute from the wrapper AFAIK.

**Comment by nagisa** [comment]
> Such as?

At the very least the alignment on the load/store instructions must not be overaligned compared to the alignment on the data allocation itself (alloca, global allocation, etc). That holds true for the regular loads/stores as well, but with the code referenced in @eddyb’s comment, we are very prone to getting ourselves into a situation where such a thing occurs in practice.

> However, using the intrinsics with the Atomic* types instead of the primitive seems sketchy, since it relies on them being represented exactly as the primitive in LLVM IR.

There was a a change recently to make these types `#[repr(transparent)]`, although that wes reverted. With this attribute it would be fine to do so, I think? 

Honestly, I do not care what exact approach is taken, honestly, as long as we can implement the atomic intrinsic without adding dubious code (use of pref alignment instead of abi alignment) into codegen. Doing so would involve carrying knowledge of alignment specified for these atomic types one way or the other.

I want at least an ICE to occur at codegen time, but I wouldn’t mind if somebody cared about this enough to also implement a proper diagnostic somewhere.

**Comment by rkruppe** [comment]
> At the very least the alignment on the load/store instructions must not be overaligned compared to the alignment on the data allocation itself (alloca, global allocation, etc).

OK that's true, but as you say it's not really a requirement of atomics. Fact is, atomic accesses require natural alignment, so the ABI alignment of the primitive type is not right. One perfectly fine to resolve that tension is to just say the `atomic_*` intrinsics require natural alignment, not just ABI alignment. Just as e.g. `_mm256_load_pd` takes an f64 pointer but requires it to be 256 bit aligned.

Using a different type whose ABI align is its size would also satisfy these constraints but has other problems, e.g.:

> There was a a change recently to make these types #[repr(transparent)], although that wes reverted. With this attribute it would be fine to do so, I think?

It was reverted specifically because of this issue (transparent is incompatible with increase alignment). If we care about supporting AtomicFoo on platforms where the primitive $Foo isn't already naturally aligned, we can't use `repr(transparent)` for atomics.

> Honestly, I do not care what exact approach is taken, honestly, as long as we can implement the atomic intrinsic without adding dubious code (use of pref alignment instead of abi alignment) into codegen.

Agreed that pref alignment is plain wrong, it's just a coincidence that it's the right alignment for all/most types and targets we care about.

**Comment by rkruppe** [comment]
I [just learned](https://godbolt.org/z/dr38BR) that LLVM does not complain about atomic loads/stores with `align` less than the size of the type. So it's *silent* UB, setting ABI alignment won't even give us an ICE.

**Comment by nagisa** [comment]
Some tidbits from an [IRC discussion](https://botbot.me/mozilla/rustc/2018-09-09/?msg=104178691&page=1):

* Unclear what alignment we document as required for `*mut T` for the `*mut T as &AtomicT` conversion to be valid:
    * Natural alignment is a conservative option (i.e. `align_of::<AtomicT> == size_of::<T>`) that makes the fewest amount of requirements possible (and also compatible with C);
    * `align_of::<AtomicT>` allows for `align_of::<AtomicT> > size_of::<T>`, if that is ever necessary for anything, but disallows underaligned pointers that could still be naturally aligned (@rkruppe points out RISCV as an example [where this could be beneficial](https://reviews.llvm.org/D47882), but also conflicts with C interop if the `AtomicT` alignment is not natural);
    * Going from requiring just "natural" alignment to `align_of::<AtomicT>` is a breaking change that would be very, very hard to make in the future.

That is all from me for the moment.

*/me makes a note to not use non-`AtomicUsize` types*

**Comment by Aaronepower** [comment]
Tirage; @cramertj What's the current status of the review of this PR?

**Comment by cramertj** [comment]
@Aaronepower https://github.com/rust-lang/rust/pull/53979 already landed removing the `#[repr(transparent)]`. I think the discussion now is just about whether or not we should provide `#[repr(transparent)]` for these types on architectures that support it.

**Comment by pnkfelix** [comment]
discussed briefly at T-compiler meeting. We cannot make a spot decision on this without more careful review. Re-assigning to @nagisa for now to make a summary comment (potentially linking to other relevant bugs) for the rest of the compiler team to read.

**Comment by nikomatsakis** [comment]
I'm a bit confused about what the points of decision are here. Can someone maybe crystallize into the key questions at hand?

My impression from @rkruppe's comment:

> Oh, I missed that LLVM mandates this. Then I guess it's settled. Note that this has more repercussions than just removing repr(transparent) – with or without the attribute, the heightened alignment requirement means that one can't safely cast any &u64 to &AtomicU64 and operate on it, unless align_of::<u64>() >= 8. Well, it's really only a problem on platforms where u64 is not naturally aligned yet AtomicU64 exists, which as I wrote in an earlier comment think is very uncommon.

Is that (a) we really just have to do whatever LLVM mandates and (b) this may mean that `&mut T -> &AtomicT` casts are not generally safe across platforms. If that's the case, can we just change the alignment and try to document the alignment requirements (and which platforms meet those restrictions by default)?

Also note that `&T -> &AtomicT` is just...never safe to do, despite the many comments writing it. I presume it is some kind of shorthand? An e.g. `&u64` is guaranteed not to change, or am I missing something.

**Comment by cramertj** [comment]
@nikomatsakis 

> Also note that &T -> &AtomicT is just...never safe to do, despite the many comments writing it. I presume it is some kind of shorthand? An e.g. &u64 is guaranteed not to change, or am I missing something.

I think you can s/`&T -> &AtomicT`/`&mut T -> &AtomicT`/g for the purposes of this conversation and the discussion all still applies.

**Comment by nagisa** [comment]
So, I was tasked with summarizing this. Here it goes (tis gonna be a long one).

First, there is one critical requirement that we cannot avoid to keep this correct:

For atomic operations to be valid on `x: *mut T`, `x.align_offset(size_of::<T>)` must be equal to `0`. In other words, the pointer on which atomic operations are done must be aligned to at least the size of `T`.

There’s no way around it, *even* if natural alignment of the type is less than its size (e.g. `i128` in Rust currently is naturally aligned to 8 bytes, on some targets `i64` may be naturally aligned to `4` bytes etc). It is user’s responsibility to ensure the correct alignment.

In C/C++ land this is solved by having types `atomic_T_t` and `std::atomic<T>` which have proper alignments and otherwise satisfy all platform-specific constraints. Since Rust:

1. Exposes a safe UB-less API for this functionality; and
2. Do not expose a way for the user to ensure these requirements are met…

The safe API *must* by all means ensure that these requirements (and all other machine-specific requirements) are met by itself. With that out of the way… What follows are some of the options:

---

We could simply "just" specify `repr(align)` and not guarantee any sort of representational equivalency between Rust’s atomics and C atomics. This would make atomic operations in FFI context roughly impossible.

This is the approach taken by this PR.

---

> Make `AtomicT` have both `repr(transparent)` and `repr(align(correct value))`

Only if we include alignment into list of type “equivalency” requirements. On one hand, that would mean that `u8`/`uint8_t` and `#[repr(transparent, align(2)] struct U8(u8)` would no longer be types with equivalent representation. On the other hand, C (gcc) provides ability to create a type equivalent to `U8` via `typedef uint8_t more_aligned_uint8_t __attribute__ ((aligned (2)));` and Rust cannot represent such type at all.

Making this (dual repr attributes) possible would solve all the problems with atomics, AFAIK.

---

> Make a way to convert from `*mut atomic_T_t` to `&AtomicT`

In order to use atomics in an FFI context it is critical to make it possible to execute atomic operations on pointers obtained over the FFI boundary. After all the whole purpose of the atomic operations is to work on areas of memory, not on values specifically.

The question then is, how do we expose atomic operations for pointers obtained via FFI? One option is to allow (the fairly unsafe) equivalence between (i.e. the proposal above):

```
/* in C */ extern void my_atomic_operation(atomic_T_t *);
/* in Rust */ extern "C" fn my_atomic_operation(&AtomicT);
```

Another option is to have simply:

```
impl AtomicT { 
    // UB if pointer is not appropriately aligned or does not satisfy any other machine-specific requirements. Doing this with `T = atomic_T_t` should always be valid.
    unsafe fn from_ptr(x: *mut T) -> &AtomicT { ... }
}
```

---

It is important to remember that the alignment for the atomic instruction itself (in generated LLVM code) must be specified and must satisfy the requirements mentioned above. This means that the way we call the atomic intrinsics (such as [this](https://doc.rust-lang.org/nightly/std/intrinsics/fn.atomic_and.html)) must change as well (i.e. instead of being called with e.g. `T = u8`, they should be called with `T = atomic_u8_t` or `T = AtomicT`).

This made me think of another option of which I haven’t thought of before, which is changing the implementation of the `Atomic` itself (for a concrete example using `AtomicU8`) from:

```rust
/* #[ ... ] */
struct AtomicU8 {
    v: UnsafeCell<u8>
}
```

to 

```rust
#[repr(align(correct))]
newtype atomic_u8_t = /*???*/;

#[repr(transparent)]
struct AtomicU8 {
    v: UnsafeCell<atomic_u8_t>
}
```

**Comment by RalfJung** [comment]
> We could simply "just" specify repr(align) and not guarantee any sort of representational equivalency between Rust’s atomics and C atomics. This would make atomic operations in FFI context roughly impossible.

Is that true? It's still a newtype, just without `repr(transparent)`. So pointers to atomics are compatible with pointers to equally-aligned C atomics (if we assume the newtype guarantees that are already relied upon in several places). Just passing them by value across FFI is not okay, but that would be a rather strange thing to do for an atomic (you want to share it by reference, not value).

Basically, I agree with @rkruppe.

**Comment by nagisa** [comment]
> Is that true? It's still a newtype, just without repr(transparent).

I do not trust `transmute::<&Y>(&x)` and `&x.y` to be equivalent, given `#[repr(Rust, align(A))] struct X { y: UnsafeCell<Y> }` and `let x; X;`. This property is necessary to make conversion from `*mut T` to `&AtomicT` possible, which by extension makes operations on FFI stuff possible at all.

I’m not sure of what newtype guarantees you’re speaking about.

**Comment by RalfJung** [comment]
> I’m not sure of what newtype guarantees you’re speaking about.

Basically, newtypes having the same layout (size and field offsets) as the type they wrap. Like, your `X` having the same layout as `UnsafeCell<Y>`. We don't guarantee this for `repr(Rust)`, but we might. If you use `repr(C)`, I think we have to guarantee it -- I assume it's what the C ABI mandates?

So just using `#[repr(C, align(A))]` should be sufficient.

**Comment by pnkfelix** [comment]
Discussed at compiler team meeting.

After reviewing @nagisa 's summary comment and the follow-on comments, we are inclined to try to experiment with using `#[repr(C, align(X))]` to resolve this.

@nagisa has agreed to take point on ushering in that change to this PR.

**Comment by nagisa** [comment]
This is the link to [the discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/subject/.2354818.20weekly.20meeting.202018-10-18/near/136048137).

**Comment by nikomatsakis** [comment]
@nagisa 

> I do not trust `transmute::<&Y>(&x)` and `&*x.y.get()` to be equivalent, given `#[repr(Rust, align(A))] struct X { y: UnsafeCell<Y> }` and `let x; X;`. 

Why not?  What are the steps required to make that true:

- `UnsafeCell<Y>` has to have the same size and alignment as `Y` (seems like that should obviously be something we guarantee)
- same with single-field structs like `struct X { y: UnsafeCell<Y> }` (I also this is pretty clearly a good idea)

Is there some other subtlety involved?

**Comment by nagisa** [comment]
@nikomatsakis my distrust with that comes specifically from `repr(Rust)` not being specified and therefore `assert_eq!(&x as *const _, &x.y as *const _)` not necessarily being true.

While I agree that “newtypes” having equivalent memory layout as the types they contain seem like a sensible idea, it is not something that we ever explicitly specified and making *unsafe* code to rely on such unwritten rule is scary.

**Comment by nikomatsakis** [comment]
OK. I feel like it is within our power, however, to reach this decision. =) (Not on this PR per se.)
